header-validator/src/validate-json.ts

```diff
@@ -114,32 +114,56 @@ function exclusive<T>(
   }
 }
 
-function string(ctx: Context, j: Json): Maybe<string> {
-  if (typeof j === 'string') {
-    return new Some(j)
+type TypeSwitch<T> = {
+  null?: CtxFunc<null, Maybe<T>>
+  boolean?: CtxFunc<boolean, Maybe<T>>
+  number?: CtxFunc<number, Maybe<T>>
+  string?: CtxFunc<string, Maybe<T>>
+  list?: CtxFunc<Json[], Maybe<T>>
+  object?: CtxFunc<JsonDict, Maybe<T>>
+}
+
+function typeSwitch<T>(ctx: Context, j: Json, ts: TypeSwitch<T>): Maybe<T> {
+  if (j === null && ts.null !== undefined) {
+    return ts.null(ctx, j)
+  }
+  if (typeof j === 'boolean' && ts.boolean !== undefined) {
+    return ts.boolean(ctx, j)
+  }
+  if (typeof j === 'number' && ts.number !== undefined) {
+    return ts.number(ctx, j)
+  }
+  if (typeof j === 'string' && ts.string !== undefined) {
+    return ts.string(ctx, j)
+  }
+  if (Array.isArray(j) && ts.list !== undefined) {
+    return ts.list(ctx, j)
   }
-  ctx.error('must be a string')
+  if (isObject(j) && ts.object !== undefined) {
+    return ts.object(ctx, j)
+  }
+
+  const allowed = Object.keys(ts)
```

Comment:
Not required yet, but do we want to add support for formatting > 2 items in the type switch now ?

Reply:
It is already supported, e.g. `a list or a string or an object`.